refactor(ivy): remove styling state storage and introduce direct style writing

matsko opened the pull request:
This patch is a final major refactor in styling Angular.

This PR includes three main fixes:
- All temporary state taht is persisted between template style/class application
  and style/class application in host bindings is now removed.
- Removes the `styling()` and `stylingApply()` instructions.
- Introduces a "direct apply" mode that is used apply prop-based
  style/class in the event that there are no map-based bindings as
  well as property collisions.

matsko commented:
Current Benchmarks:

```
[VE]
CREATE = time: 41.33235099332234 ms / iteration
CREATE_DIFF = 0%
UPDATE = time: 6.8551655627948245 ms / iteration
UPDATE_DIFF = 0%

[IVY - Master]
CREATE = time: 53.1655298012547 ms / iteration
CREATE_DIFF = 28%
UPDATE = time: 7.157582781371654 ms / iteration
UPDATE_DIFF = 4%

[IVY - THIS PR]
CREATE = time: 45.67341059578362 ms / iteration
CREATE_DIFF = 10%
UPDATE = time: 7.0579470197815874 ms / iteration
UPDATE_DIFF = 2%
```

matsko commented:
Presubmit = http://test/OCL:266449721:BASE:266449752:1567198550650:eee361a5

Ivy Presubmit = http://test/OCL:266449721:BASE:266449878:1567198589588:d30e3cd1

matsko commented:
# Benchmarks

## Large Table

### VE
CREATE = 38.435331126270256 ms
UPDATE = 6.078046357986943 ms

### IVY (based on master)
CREATE = 54.54857615973347 ms (40% slower)
UPDATE = 7.8921523179345465 ms (16% slower)

### IVY (based on this PR)
CREATE = 46.36072847700758 ms (20% slower)
UPDATE = 7.016092714709415 ms (16% slower)

## Micro Benchmarks

### [style.prop] (result = 45% faster)
MASTER = 20.81s user 0.21s system 103% cpu 20.339 total
THIS PR = 11.33s user 0.08s system 100% cpu 11.372 total

### [style.prop] and [class.prop] (result = 28% faster)
MASTER = 30.63s user 0.27s system 102% cpu 30.244 total
THIS PR = 21.80s user 0.11s system 100% cpu 21.882 total

### [style] and [class] (result = same)
MASTER = 68.91s user 1.22s system 104% cpu 1:07.15 total
THIS PR = 67.72s user 1.07s system 102% cpu 1:06.83 total

pkozlowski-opensource commented on packages/core/src/render3/instructions/select.ts:
```diff
@@ -41,6 +42,8 @@ export function selectInternal(lView: LView, index: number, checkNoChangesMode:
   ngDevMode && assertGreaterThan(index, -1, 'Invalid index');
   ngDevMode && assertDataInRange(lView, index + HEADER_OFFSET);
 
+  executeElementExitFn();
```
This is a rather expensive operation in the perf-critical path. Can you think of ways of guarding this logic so we don't need to do this call in case there is no styling?

pkozlowski-opensource commented on packages/core/src/render3/instructions/select.ts:
```diff
@@ -64,4 +67,7 @@ export function selectInternal(lView: LView, index: number, checkNoChangesMode:
   setSelectedIndex(index);
+
+  resetStylingState();
+  resetActiveElementFlags();
```
This is a rather expensive operation in the perf-critical path. Can you think of ways of guarding this logic so we don't need to do this call in case there is no styling?

pkozlowski-opensource commented on packages/core/src/render3/state.ts:
```diff
@@ -164,13 +186,45 @@ let activeDirectiveSuperClassHeight = 0;
+ *
+ * @param fn
+ */
+export function setElementExitFn(fn: Function): void {
```
So, the way I understand it is that we are setting a global state here to have a function to execute and the element "end".  Also, my understanding is that a function to execute could be figured out on the first template pass and stored on TNode.

I would suggest that we:
* introduce a new field on TView / TNode holding those functions to execute;
* guard execution of those functions with a TNode flag
* remove global state.

Obviously I'm missing toons of details here and what I'm suggesting might not be doable or trivial - at this point this is a broad-strokes suggestion, to be refined.

mhevery commented on packages/core/src/render3/instructions/select.ts:
```diff
@@ -64,4 +69,10 @@ export function selectInternal(lView: LView, index: number, checkNoChangesMode:
   // will be altered by the time we leave the `ɵɵselect` instruction.
   setSelectedIndex(index);
+
+  if (hasActiveElementFlag(ActiveElementFlags.ResetStylesOnExit)) {
```
per our discussion once flags are folded into `_selectedIndex` this reseting becomes unnecessary.

matsko replied:
Fixed.

mhevery commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -500,7 +503,9 @@ function executeTemplate<T>(
     }
     templateFn(rf, context);
   } finally {
+    executeElementExitFn();
```
```suggestion
    if (hasActiveElementFlag(ActiveElementFlags.RunExitFn)) {
      executeElementExitFn();
    }
```

matsko replied:
Fixed.

mhevery commented on packages/core/src/render3/state.ts:
```diff
@@ -129,31 +129,56 @@ export function getLView(): LView {
+ */
+export function resetActiveElementFlags() {
+  if (_selectedIndexFlags !== ActiveElementFlags.Initial) {
+    _selectedIndexFlags = ActiveElementFlags.Initial;
```
per our discussion store the state in `_selectedIndex`

matsko replied:
Fixed.

mhevery commented on packages/core/src/render3/styling_next/util.ts:
```diff
@@ -197,17 +218,22 @@ export function getStylingMapArray(value: TStylingContext | StylingMapArray | nu
+export function isStylingMapArray(value: TStylingContext | StylingMapArray | null): boolean {
+  // the StylingMapArray is in the format of [initial, prop, string, prop, string]
+  // and this is the defining value to distinguish between arrays
+  return Array.isArray(value) && isStylingContext(value) === false;
```
this code will check `isArray` twice which is not necessary.

matsko replied:
Fixed

mhevery commented on packages/core/src/render3/styling_next/util.ts:
```diff
@@ -225,6 +251,12 @@ export function getMapProp(map: StylingMapArray, index: number): string {
   return map[index + StylingMapArrayIndex.PropOffset] as string;
 }
 
+const MAP_DIRTY_VALUE = {};
```
```suggestion
const MAP_DIRTY_VALUE = typeof ngDevMode !== 'undefined' && ngDevMode ? {}: {MAP_DIRTY_VALUE:true };
```
This makes debugging easier.

matsko replied:
Updated

mhevery commented on packages/core/src/render3/styling_next/interfaces.ts:
```diff
@@ -283,94 +323,94 @@ import {LView} from '../interfaces/view';
+  HasPropAndMapBindings = 0b0000011,
+
+  /**
+   * Whether or not there are two or more sources for a single property in the context
```
It would be nice if you would include a code example demonstrating when this flag would be triggered. Otherwise "two or more source for a single property" does not mean much to a reader

(Do same for the rest of the flags)

matsko replied:
Updated

mhevery commented on packages/core/src/render3/styling_next/bindings.ts:
```diff
@@ -316,129 +326,99 @@ function allocateNewContextEntry(
 
-function contextHasUpdates(context: TStylingContext | null, bitMask: number) {
-  return context && bitMask > BIT_MASK_START_VALUE;
+  resetStylingState();
```
Is this necessary?  I would assume that the next `select` instruction would take care of the reset.

matsko replied:
Yes this is necessary. The `stylingState` has nothing to do with the `select` bits.

mhevery commented on integration/_payload-limits.json:
```diff
@@ -26,15 +26,6 @@
       }
     }
   },
-  "cli-hello-world-ivy-i18n": {
```
Why is this removed?

matsko commented on packages/core/test/bundling/hello_world/bundle.golden_symbols.json:
```diff
@@ -479,6 +491,9 @@
     "name": "syncViewWithBlueprint"
   },
+  {
+    "name": "unsetActiveElementFlag"
```
Done

matsko commented on integration/_payload-limits.json:
```diff
@@ -21,20 +21,11 @@
       "uncompressed": {
         "runtime": 1440,
-        "main": 123904,
+        "main": 125448,
```
The payload limit amounts were off (it was rebased many times) and the stylingApply code isn't gone (we still flush things like before), but the instruction is now called implicitly after one or more styling instructions is executed.

kara replied:
By gone, I meant from the generated code :-)  But yeah, that makes sense.

mhevery commented on packages/core/src/render3/state.ts:
```diff
@@ -185,26 +170,21 @@ export function unsetActiveElementFlag(flag: ActiveElementFlags) {
 }
 
 let _elementExitFn: Function|null = null;
 export function executeElementExitFn() {
```
Can we get rid of `executeElementExitFn` since all it does is just call `_elementExitFn`?

matsko replied:
No we can't because you can't export a single variable while having that variable be changed elsewhere.

matsko commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -500,6 +503,12 @@ function executeTemplate<T>(
+    if (hasActiveElementFlag(ActiveElementFlags.RunExitFn)) {
+      executeElementExitFn();
+    }
+    if (hasActiveElementFlag(ActiveElementFlags.ResetStylesOnExit)) {
```
If there is an exit function then the styles need to also be reset. The case for this is that there is a final element that runs before a templateFn finishes (no final `select()` there). 

Note that this will be removed in the next PR (there will only be an exitFn that will handle both the reset and apply).

mhevery commented on packages/core/src/sanitization/bypass.ts:
```diff
@@ -89,10 +87,10 @@ class SafeResourceUrlImpl extends SafeValueImpl implements SafeResourceUrl {
 }
 
-export function unwrapSafeValue(value: SafeValue): string {
+export function unwrapSafeValue(value: any, backupValue: string = ''): string {
```
`backupValue` is always `value`. Can we just drop the second argument?

matsko replied:
I do not understand what you mean here.

mhevery reviewed (approved):
with minor comments

matsko commented on packages/core/src/render3/styling_next/util.ts:
```diff
@@ -21,124 +44,122 @@ const TEMPLATE_DIRECTIVE_INDEX = 0;
+ * Determines whether or not to apply styles/classes directly or via context resolution.
+ */
+export function allowDirectStyling(context: TStylingContext, hostBindingsMode: boolean): boolean {
+  return isContextLocked(context, hostBindingsMode) &&
```
Yes. We should check the config value once.

matsko commented on packages/core/test/bundling/hello_world/bundle.golden_symbols.json:
```diff
@@ -420,7 +432,7 @@
   },
   {
-    "name": "resetAllStylingState"
+    "name": "resetActiveElementFlags"
```
Removed

matsko commented on packages/core/src/render3/styling_next/util.ts:
```diff
@@ -253,3 +284,134 @@ export function forceStylesAsString(styles: {[key: string]: any} | null | undefi
 }
+
+export function normalizeStylingValue(value: any) {
+  return value instanceof SafeValueImpl ? unwrapSafeValue(value) : value;
```
Updated

matsko commented on packages/core/src/render3/styling_next/util.ts:
```diff
@@ -21,124 +44,121 @@ const TEMPLATE_DIRECTIVE_INDEX = 0;
-export function setConfig(context: TStylingContext, value: number) {
+export function hasConfig(context: TStylingContext, flag: TStylingConfig, bitsMustMatch?: boolean) {
+  const value = getConfig(context) & flag;
+  return bitsMustMatch ? value === flag : value > 0;
```
Fixed

matsko committed:
refactor(ivy): remove styling state storage and introduce direct style writing

This patch is a final major refactor in styling Angular.

This PR includes three main fixes:

All temporary state taht is persisted between template style/class application
and style/class application in host bindings is now removed.
Removes the styling() and stylingApply() instructions.
Introduces a "direct apply" mode that is used apply prop-based
style/class in the event that there are no map-based bindings as
well as property collisions.

kara commented on aio/scripts/_payload-limits.json:
```diff
@@ -28,7 +28,7 @@
         "runtime-es5": 2932,
         "runtime-es2015": 2938,
-        "main-es5": 550854,
+        "main-es5": 552068,
```
Do we know what's causing the size increase?

kara replied:
@matsko and I discussed. He will look into this and fix the size regression as a follow-up PR.

kara commented on packages/core/test/render3/perf/style_binding/index.ts:
```diff
@@ -30,69 +30,39 @@ import {setupRootViewWithEmbeddedViews} from '../setup';
     ɵɵstyleProp('background-color', 'color9');
-    ɵɵstylingApply();
-    ɵɵadvance(1);
+    ɵɵadvance(10);
```
The index changes here don't look right. `advance` instructions should be relative, so this would be taking you to index ~55. Can you revert the index changes?

matsko replied:
Fixed already

kara commented on packages/core/test/render3/perf/style_and_class_bindings/index.ts:
```diff
@@ -31,89 +31,50 @@ import {setupRootViewWithEmbeddedViews} from '../setup';
     ɵɵclassProp('scale', true);
-    ɵɵstylingApply();
-    ɵɵadvance(1);
+    ɵɵadvance(10);
```
Same here - the index changes are invalid

matsko replied:
Fixed

kara commented on packages/core/test/render3/perf/map_based_style_and_class_bindings/index.ts:
```diff
@@ -30,79 +30,49 @@ import {setupRootViewWithEmbeddedViews} from '../setup';
     ɵɵclassMap('one two');
-    ɵɵstylingApply();
-    ɵɵadvance(1);
+    ɵɵadvance(10);
```
Fix here too

matsko replied:
Fixed

kara commented on packages/core/src/sanitization/bypass.ts:
```diff
@@ -89,10 +87,10 @@ class SafeResourceUrlImpl extends SafeValueImpl implements SafeResourceUrl {
 }
 
-export function unwrapSafeValue(value: SafeValue): string {
+export function unwrapSafeValue(value: any): string {
```
Curious why this is necessary

matsko replied:
Me and Misko discussed this. The end result is that the return of the empty string is the wrong behavior. It should just unwrap the value if possible

kara commented on packages/core/src/render3/state.ts:
```diff
@@ -129,31 +129,38 @@ export function getLView(): LView {
+/**
+ * Flags used for an active element during change detection.
+ *
+ * These flags are used to signal whether to run any sub routine code when
```
What does subroutine code mean here? Example would be helpful

```suggestion
 * These flags are used to signal whether to run any sub routine code when
```

matsko replied:
```
 * These flags are used within other instructions to inform cleanup or
 * exit operations to run when an element is being processed
```

kara replied:
👍

matsko committed:
fixup! refactor(ivy): remove styling state storage and introduce direct style writing

matsko committed:
fixup! refactor(ivy): remove styling state storage and introduce direct style writing

kara reviewed (approved):
LGTM
